[Concurrency] Implement basic runtime support for task futures.

DougGregor opened the pull request:
Extend AsyncTask and the concurrency runtime with basic support for
task futures. AsyncTasks with futures contain a future fragment with
information about the type produced by the future, and where the
future will put the result value or the thrown error in the initial
context.

We still don't have the ability to schedule the waiting tasks on an
executor when the future completes, so this isn't useful for anything
just test, and we can only test limited code paths.

DougGregor committed:
[Concurrency] Implement basic runtime support for task futures.

Extend AsyncTask and the concurrency runtime with basic support for
task futures. AsyncTasks with futures contain a future fragment with
information about the type produced by the future, and where the
future will put the result value or the thrown error in the initial
context.

We still don't have the ability to schedule the waiting tasks on an
executor when the future completes, so this isn't useful for anything
just test, and we can only test limited code paths.

DougGregor commented:
@swift-ci please smoke test

rjmccall commented on include/swift/ABI/Task.h:
```diff
@@ -230,23 +236,117 @@ class AsyncTask : public HeapObject, public Job {
+
+    /// Queue containing all of the tasks that are waiting in `get()`.
+    /// FIXME: do we also need a context pointer for each?
+    std::atomic<AsyncTask*> waitQueue;
```
The answer: no, waiting tasks are self-contained.

rjmccall commented on include/swift/ABI/Task.h:
```diff
@@ -230,23 +236,117 @@ class AsyncTask : public HeapObject, public Job {
+
+  private:
+    /// Status of the future.
+    std::atomic<Status> status;
```
Why is this separate from `waitQueue`?

DougGregor replied:
Oh, of course. I can absolutely unify them. Thank you!

rjmccall commented on include/swift/ABI/Task.h:
```diff
@@ -230,23 +236,117 @@ class AsyncTask : public HeapObject, public Job {
+    unsigned resultOffset;
+
+    /// The offset of the error in the initial asynchronous context.
+    unsigned errorOffset;
```
What's the purpose of these?

rjmccall commented on include/swift/ABI/Task.h:
```diff
@@ -230,23 +236,117 @@ class AsyncTask : public HeapObject, public Job {
+  ///
+  /// \returns the status of the future. If this result is
+  /// \c Executing, then \c waitingTask has been added to the
+  /// wait queue and will be scheduled when the future completes or
```
Right.

DougGregor replied:
Right. Thanks, pulled into my local tree.

rjmccall commented on include/swift/ABI/Task.h:
```diff
@@ -188,12 +190,16 @@ class AsyncTask : public HeapObject, public Job {
   void *AllocatorPrivate[4];
 
+  /// The next task in the linked list of waiting tasks.
+  AsyncTask *NextWaitingTask;
```
It's intended that `Job::SchedulerPrivate` is for things like this.

DougGregor replied:
Okay, I'm using it now.

rjmccall commented on include/swift/Runtime/Concurrency.h:
```diff
@@ -49,6 +49,33 @@ AsyncTaskAndContext swift_task_create_f(JobFlags flags,
+    size_t resultOffset, size_t errorOffset);
+
+/// Create a task object with no future which will run the given
+/// function.
```
This does have a future.

rjmccall commented on stdlib/public/Concurrency/Task.cpp:
```diff
@@ -21,20 +21,129 @@
+
+size_t FutureFragment::fragmentSize(const Metadata *resultType) {
+  return storageOffset(resultType) + resultType->vw_size();
+}
```
I'm not sure why this is out-of-line give that `storageOffset` isn't.

DougGregor replied:
No reason, fixed!

rjmccall commented on stdlib/public/Concurrency/Task.cpp:
```diff
@@ -21,20 +21,129 @@
+    break;
+
+  case Status::Error:
+    swift_unknownObjectRelease(getStoragePtr());
```
Please do not release an interior pointer into the future fragment. :)

Also, please use `swift_errorRelease`.

DougGregor replied:
`swift_errorRelease` is only defined when Objective-C interoperability is enabled. But yeah, the other thing is bad ;)

rjmccall commented on stdlib/public/Concurrency/Task.cpp:
```diff
@@ -21,20 +21,129 @@
+  while (true) {
+    waitingTask->NextWaitingTask = fragment->waitQueue.load();
+    if (fragment->waitQueue.compare_exchange_strong(
+            waitingTask->NextWaitingTask, waitingTask)) {
```
The reason you should unify the two atomic fields is that you have a race here against someone concurrently completing the task.  The thread that completes the task should exchange the appropriate completion value into the queue and then process the waiting tasks they just loaded (via the exchange).  Here you need to loop trying to add the task to the queue until you see success.

Always use memory orderings when working with `std::atomic`.  This needs to be a `std::memory_order_release` on success; it can be `std::memory_order_relaxed` on failure, except actually it needs to be `std::memory_order_acquire` because you're going to check for success/error.

Don't repeatedly reload `waitQueue`.  Load it (with an acquire) into a local, then use that local in the `compare_exchange`.  When you see a pointer, copy it into `NextWaitingTask` (or really the `SchedulerPrivate` contents).

You should use `compare_exchange_weak` since you're already doing this in a loop.

DougGregor replied:
Thank you! I've made all of these changes together. It looks much more reasonable now.

rjmccall commented on stdlib/public/Concurrency/Task.cpp:
```diff
@@ -21,20 +21,129 @@
+  auto resultPtr = reinterpret_cast<OpaqueValue *>(
+      reinterpret_cast<char *>(context) + fragment->resultOffset);
+  fragment->resultType->vw_initializeWithTake(storagePtr, resultPtr);
+  fragment->status = FutureFragment::Status::Success;
```
The general ABI for indirect return values is that the address to fill would be passed as an argument, not that we'd make space for it in the initial context and that move out of that into the future.

I don't remember what convention we decided on for errors, but I doubt this it's this strange.  Should there be a flag in the task reporting whether it can throw?

rjmccall commented on stdlib/public/Concurrency/Task.cpp:
```diff
@@ -62,10 +171,17 @@ static void completeTask(AsyncTask *task, ExecutorRef executor,
+    task->completeFuture(context);
+
+    // Schedule tasks that are waiting on the future.
+    task->scheduleWaitingTasks(executor);
```
As mentioned, these need to be combined so that you can deal with the queue atomically.

DougGregor replied:
Done, thanks!

rjmccall commented on stdlib/public/Concurrency/Task.cpp:
```diff
@@ -21,20 +21,129 @@
+
+  // FIXME: "Simple task" seems like it's not that useful a concept. We might
+  // also have task-local storage to destroy.
+  // assert(!task->isFuture());
```
I agree that it probably makes sense to just have the one task-destruction function.  You can just rename this.

DougGregor committed:
[Future] Add a test to make sure we're handling object lifetimes properly.

DougGregor commented on include/swift/ABI/Task.h:
```diff
@@ -230,23 +236,117 @@ class AsyncTask : public HeapObject, public Job {
+          reinterpret_cast<char *>(this) + storageOffset(resultType));
+    }
+
+    /// Retrieve a reference to the storage of the 
```
It's a stray comment from a method that was removed, I'll zap it. Thanks!

DougGregor commented on include/swift/ABI/Task.h:
```diff
@@ -230,23 +236,117 @@ class AsyncTask : public HeapObject, public Job {
+  /// wait queue and will be scheduled when the future completes or
+  /// is cancelled. Otherwise, the future has completed and can be
+  /// queried.
+  FutureFragment::Status waitFuture(AsyncTask *waitingTask);
```
If the future is still executing, it puts `waitingTask` on the queue to be scheduled once the future has completed. We then get to yield the thread back to the scheduler to do more work (but don't have the way to do that yet)

DougGregor commented on stdlib/public/Concurrency/Task.cpp:
```diff
@@ -21,20 +21,129 @@
+  auto storagePtr = fragment->getStoragePtr();
+
+  // Check for an error.
+  if (unsigned errorOffset = fragment->errorOffset) {
```
It could be zero for a non-throwing function. We should probably turn that into a `JobFlag`.

ktoso replied:
Ah I see, thanks! Yeah may be a job flag would be nicer.

DougGregor committed:
[Futures] Fix up some comments to address some comments.

DougGregor committed:
[Concurrency] Use a single atomic for future wait queue.

Use a single atomic for the wait queue that combines the status with
the first task in the queue. Address race conditions in waiting and
completing the future.

Thanks to John for setting the direction here for me.

DougGregor committed:
[Concurrency] Use SchedulerPrivate for the "next waiting task" link.

Reduce the size of AsyncTask by using the first slot of SchedulerPrivate
for the next waiting task. Thanks, John!

DougGregor committed:
[Concurrency] Turn "simple task" into just "task".

There isn't a big advantage to keeping "simple" task as a separate
concept from a task. Tasks may or may not have futures.

DougGregor committed:
[Concurrency] Inline AsyncTask::FutureFragment::fragmentSize.

DougGregor committed:
[Concurrency] Clarify/fix error object handling in futures.

DougGregor committed:
[Concurrency] Have future functions write their results directly.

Introduce `FutureAsyncContext` to line up with the async context formed
by IR generation for the type `<T> () async throws -> T`. When allocating
a future task, set up the context with the address of the future's storage
for the successful result and null out the error result, so the caller
will directly fill in the result. This eliminates a bunch of extra
complexity and a copy.

rjmccall commented on stdlib/public/Concurrency/Task.cpp:
```diff
@@ -21,51 +21,164 @@
+      newQueueHead, std::memory_order_acquire);
+  assert(queueHead.getStatus() == Status::Executing);
+
+  // Notify every
```
Incomplete

rjmccall commented on include/swift/ABI/Task.h:
```diff
@@ -230,15 +239,137 @@ class AsyncTask : public HeapObject, public Job {
+  AsyncTask *&getNextWaitingTask() {
+    return reinterpret_cast<AsyncTask *&>(
+        SchedulerPrivate[NextWaitingTaskIndex]);
+  }
```
I think the method name here should probably reinforce that this is implementation-private in some way.

rjmccall reviewed (commented):
Thanks, looking a lot better.

DougGregor committed:
[Concurrency] Minor cleanups to futures.

DougGregor commented:
@swift-ci please test

DougGregor committed:
[Concurrency] Remove an unnecessary 'const' from a non-static data member.

'const' breaks compilation on newer compilers.

DougGregor commented:
@swift-ci please smoke test

rjmccall commented on include/swift/ABI/Task.h:
```diff
@@ -230,19 +240,141 @@ class AsyncTask : public HeapObject, public Job {
+    /// fragment.
+    static size_t storageOffset(const Metadata *resultType)  {
+      size_t offset = sizeof(FutureFragment);
+      size_t alignment = std::max(resultType->vw_alignment(), alignof(void *));
```
What is this `max` about?  The error value?  Could you use `SwiftError*` to communicate that, here and below?

...should I overlook that this is subtly pessimal in really implausible cases, like a highly-aligned byte?  I'm not sure it can practically happen.

DougGregor replied:
Yes, it's the `SwiftError*`. I can use that.

rjmccall commented on include/swift/ABI/Task.h:
```diff
@@ -230,19 +240,141 @@ class AsyncTask : public HeapObject, public Job {
+    ///
+    /// The low bits contain the status, the rest of the pointer is the
+    /// AsyncTask.
+    std::atomic<WaitQueueItem> waitQueue;
```
Technically I don't think that tasks need to coincide with any status but `Executing`, so the other statuses could be signal values rather than masked on.  I don't think it really matters, though.

rjmccall commented on stdlib/public/Concurrency/Task.cpp:
```diff
@@ -21,51 +21,150 @@
+    break;
+
+  case Status::Error:
+    swift_unknownObjectRelease(reinterpret_cast<OpaqueValue *>(getError()));
```
Is this the standard way we release errors in the runtime?  Shouldn't we have an `swift_errorRelease` we can unconditionally call, even if it just compiles to `swift_release` in some configurations?  Is it not important that we call `swift_errorRelease` in the configurations that provide it?

DougGregor replied:
This is the standard way when we don't know whether we'll have a native Swift object or not. `swift_errorRelease` is implemented as `objc_release` where it's implemented.

rjmccall commented on stdlib/public/Concurrency/Task.cpp:
```diff
@@ -21,51 +21,150 @@
+    auto nextWaitingTask = waitingTask->getNextWaitingTask();
+
+    // Remove this task from the list.
+    waitingTask->getNextWaitingTask() = nullptr;
```
You actually don't have to do this; the memory is invalidated once it's not waiting.

DougGregor replied:
Alright.

rjmccall commented on stdlib/public/Concurrency/Task.cpp:
```diff
@@ -83,13 +182,35 @@ AsyncTaskAndContext
+    JobFlags flags, AsyncTask *parent, const Metadata *futureResultType,
+    AsyncFunctionType<void()> *function, size_t initialContextSize) {
+  assert((futureResultType != nullptr) == flags.task_isFuture());
+  assert((futureResultType != nullptr) == flags.task_isFuture());
```
Seeing undouble?

Should we assert here that `initialContextSize >= sizeof(FutureAsyncContext)`?

DougGregor replied:
That's a good idea!

rjmccall commented on include/swift/ABI/Task.h:
```diff
@@ -327,6 +459,19 @@ class YieldingAsyncContext : public AsyncContext {
+/// task.
+///
+/// This type matches the ABI of a function `<T> () async throws -> T`, which
+/// is used to describe futures.
```
To be precise: it's the signature of the function taken by `runDetached` and `Task.Group.add`.  We basically fully set up the call except for any formal arguments (e.g. closure-context arguments).

DougGregor replied:
Sure.

DougGregor commented:
@swift-ci please smoke test

DougGregor committed:
[Concurrency] More cleanups for futures

DougGregor commented:
@swift-ci please smoke test

DougGregor commented:
This makes no sense...
```
FAILED: unittests/runtime/SwiftRuntimeTests.exe 
cmd.exe /C "cd . && "C:\Program Files\CMake\bin\cmake.exe" -E vs_link_exe --intdir=unittests\runtime\CMakeFiles\SwiftRuntimeTests.dir --rc=C:\PROGRA~2\WI3CF2~1\10\bin\100183~1.0\x64\rc.exe --mt=C:\PROGRA~2\WI3CF2~1\10\bin\100183~1.0\x64\mt.exe --manifests  -- C:\PROGRA~2\MICROS~1\2019\COMMUN~1\VC\Tools\MSVC\1424~1.283\bin\Hostx64\x64\link.exe /nologo @CMakeFiles\SwiftRuntimeTests.rsp  /out:unittests\runtime\SwiftRuntimeTests.exe /implib:unittests\runtime\SwiftRuntimeTests.lib /pdb:unittests\runtime\SwiftRuntimeTests.pdb /version:0.0  /INCREMENTAL:NO /STACK:10000000 /INCREMENTAL:NO /subsystem:console  && cd ."
LINK: command "C:\PROGRA~2\MICROS~1\2019\COMMUN~1\VC\Tools\MSVC\1424~1.283\bin\Hostx64\x64\link.exe /nologo @CMakeFiles\SwiftRuntimeTests.rsp /out:unittests\runtime\SwiftRuntimeTests.exe /implib:unittests\runtime\SwiftRuntimeTests.lib /pdb:unittests\runtime\SwiftRuntimeTests.pdb /version:0.0 /INCREMENTAL:NO /STACK:10000000 /INCREMENTAL:NO /subsystem:console /MANIFEST /MANIFESTFILE:unittests\runtime\SwiftRuntimeTests.exe.manifest" failed (exit code 1120) with the following output:
   Creating library unittests\runtime\SwiftRuntimeTests.lib and object unittests\runtime\SwiftRuntimeTests.exp
TaskFuture.cpp.obj : error LNK2019: unresolved external symbol "struct swift::FullMetadata<struct swift::TargetOpaqueMetadata<struct swift::InProcess> > const $sSiN" (?$sSiN@@3U?$FullMetadata@U?$TargetOpaqueMetadata@UInProcess@swift@@@swift@@@swift@@B) referenced in function "private: virtual void __cdecl TaskFutureTest_intFuture_Test::TestBody(void)" (?TestBody@TaskFutureTest_intFuture_Test@@EEAAXXZ)
unittests\runtime\SwiftRuntimeTests.exe : fatal error LNK1120: 1 unresolved externals
```

DougGregor commented:
@swift-ci please smoke test Linux

DougGregor committed:
Remove test that depends on Swift.Int metadata, which isn't around.

The objectFuture test is better anyway.

DougGregor commented:
@swift-ci please test

DougGregor commented:
Last commit should clear up the Windows failure. Rerunning all of the tests...

DougGregor commented:
@swift-ci pleas test

DougGregor commented:
Got tripped up by https://github.com/apple/swift/pull/34748, but otherwise everything was fine.

DougGregor commented:
@swift-ci please smoke test
